--- nginx-1.5.5/src/http/modules/ngx_http_ssl_module.c	2013-09-17 17:31:01.000000000 +0400
+++ nginx-1.5.5/src/http/modules/ngx_http_ssl_module.c	2013-09-28 04:02:07.086055280 +0400
@@ -160,6 +160,13 @@
       offsetof(ngx_http_ssl_srv_conf_t, session_timeout),
       NULL },
 
+    { ngx_string("ssl_session_ticket_key"),
+      NGX_HTTP_MAIN_CONF|NGX_HTTP_SRV_CONF|NGX_CONF_TAKE1,
+      ngx_conf_set_str_slot,
+      NGX_HTTP_SRV_CONF_OFFSET,
+      offsetof(ngx_http_ssl_srv_conf_t, session_ticket_keyfile),
+      NULL },
+
     { ngx_string("ssl_crl"),
       NGX_HTTP_MAIN_CONF|NGX_HTTP_SRV_CONF|NGX_CONF_TAKE1,
       ngx_conf_set_str_slot,
@@ -413,6 +420,7 @@
      *     sscf->shm_zone = NULL;
      *     sscf->stapling_file = { 0, NULL };
      *     sscf->stapling_responder = { 0, NULL };
+     *     sscf->session_ticket_keyfile = { 0, NULL };
      */
 
     sscf->enable = NGX_CONF_UNSET;
@@ -469,6 +477,10 @@
                          "");
     ngx_conf_merge_str_value(conf->trusted_certificate,
                          prev->trusted_certificate, "");
+
+    ngx_conf_merge_str_value(conf->session_ticket_keyfile,
+                        prev->session_ticket_keyfile, "");
+
     ngx_conf_merge_str_value(conf->crl, prev->crl, "");
 
     ngx_conf_merge_str_value(conf->ecdh_curve, prev->ecdh_curve,
@@ -620,6 +632,13 @@
         != NGX_OK)
     {
         return NGX_CONF_ERROR;
+    }
+
+    if (ngx_ssl_session_ticket_key(cf, &conf->ssl,
+                                   &conf->session_ticket_keyfile)
+        != NGX_OK)
+    {
+        return NGX_CONF_ERROR;
     }
 
     if (conf->stapling) {
