--- nginx-1.5.5/src/http/modules/ngx_http_ssl_module.h	2013-09-17 17:31:01.000000000 +0400
+++ nginx-1.5.5/src/http/modules/ngx_http_ssl_module.h	2013-09-28 04:11:39.698856421 +0400
@@ -30,6 +30,8 @@
 
     time_t                          session_timeout;
 
+    ngx_str_t                       session_ticket_keyfile;
+
     ngx_str_t                       certificate;
     ngx_str_t                       certificate_key;
     ngx_str_t                       dhparam;
